packages/theme/src/cli/utilities/theme-environment/local-assets.ts

```diff
@@ -94,35 +94,143 @@ function isCompiledAssetRequest(event: H3Event) {
+  })
+}
 
-    const stylesheets = ['/*** GENERATED LOCALLY ***/\n']
-    for (const [, file] of [...sectionFiles, ...blockFiles, ...snippetFiles]) {
-      const stylesheet = getCompiledStylesheet(file)
-      if (stylesheet) stylesheets.push(stylesheet)
+function handleBlockScriptsJs(ctx: DevServerContext, event: H3Event, kind: 'block' | 'snippet' | 'section') {
+  const blockFiles = getLiquidFilesByKind(ctx, kind)
+  const kinds = `${kind}s`
+
+  const javascripts = [
+    `
+      /*** GENERATED LOCALLY ***/
+
+      (function () {
+        var __${kinds}__ = {};
+
+        (function () {
+          var element = document.getElementById("${kinds}-script");
+          var attribute = element ? element.getAttribute("data-${kinds}") : "";
+          var ${kinds} = attribute.split(",").filter(Boolean);
+
+          for (var i = 0; i < ${kinds}.length; i++) {
+            __${kinds}__[${kinds}[i]] = true;
+          }
+        })();
+    `,
+  ]
+
+  for (const [key, file] of blockFiles) {
+    const baseName = key.split('/').pop()?.replace('.liquid', '')
+    const javascript = getCompiledJavaScript(file)
+
+    if (baseName && javascript) {
+      /**
+       * For sections, this approach isn't completely accurate, as we'd have
+       * something like this:
+       * ```
+       *    if (!__sections__['header'] && !window.DesignMode) return;
+       * ```
+       * However, this works well enough for most cases. It's preferable to have
+       * a consistent JavaScript workflow across sections, blocks, and snippets,
+       * especially when considering hot reloading. This only affects JavaScript
+       * sections in the theme editor and doesn't impact the main runtime.
+       */
+      javascripts.push(`
+        (function () {
+          if (!__${kinds}__["${baseName}"] && !Shopify.designMode) return;
+          try {
+            ${javascript}
+          } catch (e) {
+            console.error(e);
+          }
+        })();
+      `)
     }
+  }
 
-    const stylesheet = stylesheets.join('\n')
+  javascripts.push('})();')
 
-    return serveStatic(event, {
-      getContents: () => stylesheet,
-      getMeta: () => ({type: 'text/css', size: stylesheet.length, mtime: new Date()}),
+  const javascript = javascripts.join('\n')
+
+  return serveStatic(event, {
+    getContents: () => javascript,
+    getMeta: () => ({type: 'text/javascript', size: javascript.length, mtime: new Date()}),
+  })
+}
+
+function getLiquidFilesByKind(ctx: DevServerContext, kind: 'section' | 'block' | 'snippet') {
+  return [...ctx.localThemeFileSystem.files.entries()]
+    .filter(([key]) => {
+      return key.endsWith('.liquid') && key.startsWith(`${kind}s/`)
+    })
+    .sort(([key1], [key2]) => {
+      return key1.localeCompare(key2)
     })
-  }
 }
 
-const compiledStylesheets = new Map<string, {checksum: string; content: string}>()
+function getCompiledJavaScript(file: ThemeAsset) {
```

Comment:
Not a big issue but since this function is virtually the same as `getCompiledStylesheet`, perhaps we can extract it in `getCompiledContent(file: ThemeAsset, tag: 'javascript' | 'stylesheet', cache: Map<string, {checksum: string, content: string}>)`?

Reply:
Oops! Excellent catch! I've applied that :)